### ISSUE TITLE
"Sqlite3.dll was not loaded" error when using sqlite.raw in Visual Studio

### ISSUE THREAD

**Issue by AArnott** [ericsink/SQLitePCL.raw#54]
I have a Visual Studio VSIX that carries SQLite with it. When I try to use it outside VS it works fine. But inside VS, the exception below is always thrown. I have confirmed that these files are available in the expanded VSIX (where the rest of my code is running):
SQLite-net.dll
sqlitepcl.raw.dll
x86\sqlite3.dll
x64\sqlite3.dll

> System.TypeInitializationException occurred
  HResult=-2146233036
  Message=The type initializer for 'NativeMethods' threw an exception.
  Source=SQLitePCL.raw
  TypeName=NativeMethods
  StackTrace:
       at SQLitePCL.SQLite3Provider.NativeMethods.sqlite3_open_v2(Byte[] filename, IntPtr& db, Int32 flags, Byte[] vfs)
       at SQLitePCL.SQLite3Provider.SQLitePCL.ISQLite3Provider.sqlite3_open_v2(String filename, IntPtr& db, Int32 flags, String vfs)
       at SQLitePCL.raw.sqlite3_open_v2(String filename, sqlite3& db, Int32 flags, String vfs)
       at SQLite.SQLiteConnection..ctor(String databasePath, SQLiteOpenFlags openFlags, Boolean storeDateTimeAsTicks)
       at SQLite.SQLiteConnection..ctor(String databasePath, Boolean storeDateTimeAsTicks)
       at Microsoft.VisualStudio.*(String databasePath)
  InnerException: 
       HResult=-2146233088
       Message=sqlite3.dll was not loaded.
       Source=SQLitePCL.raw
       StackTrace:
            at SQLitePCL.SQLite3Provider.NativeMethods..cctor()
       InnerException:

**Comment by ericsink** [comment]
Usually this error comes from not being able to find the C runtime lib.  For example, if you project targets any of the RT-flavored platforms, you'll need to add a reference to the Visual C++ runtime extension SDK.

**Comment by AArnott** [comment]
Thanks. This is VS so it's Win32 (on Windows 10, in this case). I wonder if it's that the algorithm for finding sqlite3.dll isn't prepared for it to be under an x86 directory that isn't under the .exe's own directory. Is there a way for me to explicitly specify the path?

**Comment by ericsink** [comment]
Hmmm.  Maybe you're hitting something similar to issue #50.

**Comment by AArnott** [comment]
I just tried upgrading to 0.8.1, but that makes SQLitePCL.raw.dll not even deploy into my VSIX. I'm trying to figure out why that is now, but if you have any ideas that could save me time, please share.

**Comment by ericsink** [comment]
For your situation, I don't know what could be causing that.

I assume your VSIX is ending up with net45 as the chosen platform.  In that case, the platform assembly is in the nupkg's lib directory, so everything is simple.

(For some other platforms, 0.8.x moves the platform assembly into the build directory and relies on an msbuild .targets file to add the appropriate reference.  This problem should not be happening in your case.)

I do believe that 0.8.1 does contain a net45.targets file that was not there in 0.7.x, but all it does is copy the relevant sqlite3.dll into the build directory.

All that said, AFAIK you are the first person to use SQLitePCL.raw in a VSIX, so I tend to assume the problems you are encountering are new.  I'm interested in fixing obstacles for this use case.

**Comment by AArnott** [comment]
Interesting.  There was definitely a difference between 0.7.1 and 0.8.1 in whether it copies, but I wonder if it had to do with the fact that I also have SQLite-net-pcl 1.0.11, which only references 0.7.1. It seems OK with my using 0.8.1, but perhaps the version mismatch is to blame for the VSIX embedding failing.

However you said something that made me slap my forehead: I should install both SQLite-related PCL packages into the VSIX project itself rather than just in projects that the VSIX references. Once I did that, the files were all back into the VSIX the way they should.

And with that, your 0.8.1 changes, and an additional bit of code I added to load sqlite3.dll from the x86 directory deployed by the VSIX, we're all up and running. Thanks.
